designs/0032-cmdstan-output-formats.md

```diff
@@ -0,0 +1,230 @@
+The Stan model parameter inputs and outputs are on the *constrained* scale, therefore the number of parameter values reported by the `write_array` method
+is reported by the `get_constrained_sizedtypes` method.
+The sampler computes on the *unconstrained* scale, therefore the size of the metric (inverse mass matrix) is reported by the `get_unconstrained_sizedtypes` method.
+(Note: A further complication for optimization is that the optimization algorithm computes on the unconstrained scale, which means that the Hessian of the covariance matrix is on the unconstrained scale.)
+
+
+### Inference algorithm outputs
+
+The outputs from the inference algorithm necessarily differ according to the method, and to a lesser degree, according to the developers.
+The NUTS-HMC sampler does full Bayesian inference by assembling a set of draws from the posterior distribution using an HMC sampler
+with tuning parameters step-size and metric.
+The `stan::services::util::mcmc_writer` class is constructed with a callback writer and provides methods
+`write_sample_names`, `write_sample_params`, `write_adapt_finish`, `write_timing`.
+
+- The `write_sample_names` and `write_sample_params` produce the CSV header row and data rows, respectively.
+The initial columns consist of sampler state and diagnostic information from the `stan::mcmc::sample` object,
+(called `sample_params`), the remaining columns are the Stan model variable names and values returned by
+the `constrained_param_names` and `write_array` methods.
+
+- The `write_adapt_finish` and `write_timing` produce comment blocks in the CSV file.
+Adaptation information consists of the stepsize and metric.  This comment block follows
+the CSV header row, either immediately, or following the warmup data rows.
+The timing information comments are written out at the end.
+
+
+### Interface-level outputs
+
+The CmdStan interface dumps all configuration information into the Stan CSV file as a series of initial comment lines.
+The CmdStan argument parser uses indentation to indicate nested argument structure.
+Following the argument parser are several more comment lines of key-value pairs of information
+about the software version and compiler options used to compile the Stan model executable.
+
+
+## Functional specification
+
+We propose to break down the information in the current Stan CSV file into several files
+according to information source, whose format matches the information structure of the contents.
+The default file format for tabular data will be CSV.
+The default file format for hierarchical data will be JSON.
+
+Tabular data:
+
+- The Stan model variables,
+i.e., the `constrained_param_names` method generates the output column names
+and the `write_array` method generates the output data.
+
+- The inference method variables (`sampler_params` for HMC, tbd for optimization, variational inference).
+
+- The HMC metric (?).  The HMC metric is either a matrix diagonal vector or a dense matrix.
```

Comment:
The metric is a positive-definite matrix, which can be either diagonal or dense.  If it's diagonal, then we only need to store the vector corresponding to the diagonal entries.

Reply:
revised.